tiledb/common/arithmetic.h

```diff
@@ -0,0 +1,251 @@
+/**
+ * @file   arithmetic.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2025 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines common non-primitive arithmetic functions.
+ */
+#ifndef TILEDB_ARITHMETIC_H
+#define TILEDB_ARITHMETIC_H
+
+#include "tiledb/common/unreachable.h"
+
+#include <cstdint>
+#include <limits>
+#include <optional>
+
+namespace tiledb::common {
+
+/**
+ * Provides the member typedef `type` which is the 64-bit integral type
+ * with the same signed-ness as the integral type `T`,
+ * if `T` itself is narrower than 64 bits.
+ */
+template <typename T>
+struct extended_integral_type;
+
+template <>
+struct extended_integral_type<int8_t> {
+  using type = int64_t;
+};
+
+template <>
+struct extended_integral_type<int16_t> {
+  using type = int64_t;
+};
+
+template <>
+struct extended_integral_type<int32_t> {
+  using type = int64_t;
+};
+
+template <>
+struct extended_integral_type<uint8_t> {
+  using type = uint64_t;
+};
+
+template <>
+struct extended_integral_type<uint16_t> {
+  using type = uint64_t;
+};
+
+template <>
+struct extended_integral_type<uint32_t> {
+  using type = uint64_t;
+};
+
```

Comment:
```suggestion

template <class T> using extended_integral_type_t = typename extended_integral_type<T>::type;

```

Using `extended_integral_type_t` will slightly shorten some code below.

Reply:
756c11d